panic: send on close channel - during 'dogestry push'

This could be user error, but my guess is there's at least a problem with some error handling code.  

version:

```
[root@localhost conf]# dogestry version
Dogestry 2.0.1

[root@localhost conf]# uname -a
Linux localhost.localdomain 3.10.0-123.20.1.el7.x86_64 #1 SMP Thu Jan 29 18:05:33 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

command:

`dogestry push s3://mybucket-docker?region=us-west-2 someuser/register`

error:

```
Pushing files to S3 remote:
2015/07/07 14:59:59 error when uploading to S3: 307: "Please re-send this request to the specified temporary endpoint. Continue to use the original request endpoint for future requests."
{"Status":"error", "Message": "Error when uploading to S3: 307: "Please re-send this request to the specified temporary endpoint. Continue to use the original request endpoint for future requests.""}
panic: send on closed channel

goroutine 131 [running]:
github.com/dogestry/dogestry/remote.func·002()
	/Users/didip/go/src/github.com/dogestry/dogestry/remote/s3.go:137 +0x24e
created by github.com/dogestry/dogestry/remote.(*S3Remote).Push
	/Users/didip/go/src/github.com/dogestry/dogestry/remote/s3.go:143 +0x499
```

--- a/remote/s3.go
+++ b/remote/s3.go
@@ -126,7 +126,7 @@ func (remote *S3Remote) Push(image, imageRoot string) error {
 	defer close(putFileErrChan)
 
 	numGoroutines := 25
-	goroutineQuitChans := make([]chan bool, 25)
+	goroutineQuitChans := make([]chan bool, numGoroutines)
 	for i := 0; i < numGoroutines; i++ {
 		goroutineQuitChans[i] = make(chan bool)
 	}
@@ -155,6 +155,11 @@ func (remote *S3Remote) Push(image, imageRoot string) error {
 	for i := 0; i < len(keysToPush); i++ {
 		p := <-putFileErrChan
 		if p.err != nil {
+			// Close all running goroutines
+			for i := 0; i < numGoroutines; i++ {
+				goroutineQuitChans[i] <- true
+			}
+
 			log.Printf("error when uploading to S3: %v", p.err)
 			return fmt.Errorf("Error when uploading to S3: %v", p.err)
 		}

--- a/remote/s3.go
+++ b/remote/s3.go
@@ -126,21 +126,30 @@ func (remote *S3Remote) Push(image, imageRoot string) error {
 	defer close(putFileErrChan)
 
 	numGoroutines := 25
+	goroutineQuitChans := make([]chan bool, 25)
+	for i := 0; i < numGoroutines; i++ {
+		goroutineQuitChans[i] = make(chan bool)
+	}
 
 	println("Pushing files to S3 remote:")
 	for i := 0; i < numGoroutines; i++ {
-		go func() {
-			for putFile := range putFilesChan {
-				putFileErr := remote.putFile(putFile.KeyDef.fullPath, &putFile.KeyDef)
+		go func(i int) {
+			select {
+			case <-goroutineQuitChans[i]:
+				return
+			default:
+				for putFile := range putFilesChan {
+					putFileErr := remote.putFile(putFile.KeyDef.fullPath, &putFile.KeyDef)
 
-				if (putFileErr != nil) && ((putFileErr != io.EOF) && (!strings.Contains(putFileErr.Error(), "EOF"))) {
-					putFileErrChan <- putFileResult{putFile.Key, putFileErr}
-					return
+					if (putFileErr != nil) && ((putFileErr != io.EOF) && (!strings.Contains(putFileErr.Error(), "EOF"))) {
+						putFileErrChan <- putFileResult{putFile.Key, putFileErr}
+						return
+					}
+
+					putFileErrChan <- putFileResult{}
 				}
-
-				putFileErrChan <- putFileResult{}
 			}
-		}()
+		}(i)
 	}
 
 	for i := 0; i < len(keysToPush); i++ {

